### PR TITLE
ENH: add ndenumerate specialization for masked arrays

### PR DESCRIPTION
**Pull request by joukewitteveen**
Using `ndenumerate` on a masked array does not yield `ma.masked` for masked values, but rather the actual value of the underlying data array. This makes sense from a performance perspective, but is potentially undesirable. Rather than filling the masked array before using `ndenumerate` and incurring a copy, we can use a simple `ndenumerate` wrapper specifically for masked arrays. This MR suggests such a specialization of `ndenumerate`.

An 'upcoming changes' message is still missing as this is this functionality is proposed without prior discussion and I want to know what people think first.

At least in 2011 I could find someone else wanting this functionality too: https://stackoverflow.com/questions/8620798/numpy-ndenumerate-for-masked-arrays/8621173.

### CODE REVIEW

File: numpy/ma/__init__.pyi [hunk 1]
```diff
@@ -218,6 +218,7 @@ from numpy.ma.extras import (
     masked_all_like as masked_all_like,
     median as median,
     mr_ as mr_,
+    ndenumerate as ndenumerate,
```
Comment (BvB93) [inline numpy/ma/__init__.pyi:221]: FYI, you'll also have to add `ndenumerate` to the [extras.pyi](https://github.com/numpy/numpy/blob/c92864091e5928d92bc109d1505febe35f3909f1/numpy/ma/extras.pyi) stub file for this to work:
``` diff
+ from numpy.typing import ArrayLike, _Shape
+ from typing import Generator, Tuple
+
+ def ndenumerate(a: ArrayLike) -> Generator[Tuple[_Shape, Any], None, None]: ...
```

**Review by BvB93 — COMMENTED** [review]
The `.pyi` addition looks good to me.
As far as the general concept is concerned: I would like to see the opinion of someone more involved with masked arrays.

**Review by BvB93 — COMMENTED** [review]
One thing that is most definitely missing from this PR is a set of tests.
Three useful test cases would be masked arrays with all-`False`, all-`True` and mixed `True`/`False` masks.

**Review by BvB93 — APPROVED** [review]
> RuntimeError: Missing `SVML` submodule! Run `git submodule update --init` to fix this.

The CircleCI error seems to be unrelated, though I'm wondering if it is a fluke or not.

In any case, I don't have any particularly strong feelings about masked arrays, but the new `np.ma.ndenumerate` implementation looks solid enough.

**Comment by mattip** [comment]
I think we should put this in. If it exists on ndarray, it should exist on ma.

File: numpy/ma/extras.py [hunk 2]
```diff
@@ -1519,6 +1519,51 @@ def __init__(self):
 #---- Find unmasked data ---
 #####--------------------------------------------------------------------------
 
+def ndenumerate(a):
+    """
+    Multidimensional index iterator.
+
+    Return an iterator yielding pairs of array coordinates and values of
+    elements that are not masked.
+
```
Comment (mattip) [inline numpy/ma/extras.py:1528]:
```suggestion
    .. versionadded:: 1.10.0
    
```

**Review by mattip — APPROVED** [review]

**Review by rashedaoversea — APPROVED** [review]

**Comment by mattip** [comment]
From the mailing list discussion:
>> This version skips all masked elements.  An alternative could be to
>> return `np.ma.masked` for masked elements?
>
> Would it be a bad idea to add a kwarg that specifies this behaviour (i.e. offering both alternatives)? Assuming people might need the masked items to be there under certain circumstances. Perhaps when zipping masked data with dense data?

**Comment by joukewitteveen** [comment]
I added a commit on top of this MR with the additional kwarg. I hope this is in line with expectations.

**Comment by mattip** [comment]
Was there consensus around the `compressed=True` kwarg? Isn't `np.ma.ndenumerate(a, compressed=False)` equivalent to `np.ndenumerate(a)` ? If so maybe we could drop the kwarg and add a documentation note to use `np.ndenumerate()` to get that behavior

**Comment by joukewitteveen** [comment]
> Was there consensus around the `compressed=True` kwarg? Isn't `np.ma.ndenumerate(a, compressed=False)` equivalent to `np.ndenumerate(a)` ? If so maybe we could drop the kwarg and add a documentation note to use `np.ndenumerate()` to get that behavior

There was no consensus, just a request for the functionality. One of the key reasons to have the specialization is that with `np.ndenumerate` you'd get `element == a.data[i]`, whereas with `np.ma.ndenumerate` you'd get `element == a[i]`. This makes a difference for masked elements, where only in the second case we'd have `element == ma.masked`.

File: numpy/ma/extras.py [hunk 3]
```diff
@@ -1519,6 +1519,62 @@ def __init__(self):
...
+    (1, 1) 4
+    (2, 0) 6
+    (2, 2) 8
+
+    >>> for index, x in np.ma.ndenumerate(a[1], compressed=False):
+    ...     print(index, x)
+    (0,) --
+    (1,) 4
+    (2,) --
+    """
```
Comment (mattip) [inline numpy/ma/extras.py:1570]: Can you add what happens when doing `np.ma.ndenumerate(a, compressed=False)` and also `np.ndenumerate(a)`, with appropriate text describing what is going on

**Comment by seberg** [comment]
We discussed this briefly in the triage meeting, but did not want to decide on this.  Maybe we should ping the mailing list once more on the exact addition (since it is new API).  Not that I particularly expect a lot of input.

In general, the PR seems pretty much ready, and we should just make a decision to merge or not soon, though.

**Comment by charris** [comment]
This sounds reasonable to me. @mattip Any reservations?

File: numpy/ma/extras.py [hunk 4]
```diff
@@ -1520,6 +1520,68 @@ def __init__(self):
...
+    >>> for index, x in np.ma.ndenumerate(a):
+    ...     print(index, x)
+    (0, 0) 0
+    (0, 1) 1
+    (0, 2) 2
+    (1, 1) 4
+    (2, 0) 6
+    (2, 2) 8
+
+    >>> for index, x in np.ma.ndenumerate(a[1], compressed=False):
```
Comment (mattip) [inline numpy/ma/extras.py:1572]: This should be just like the previous example, only with `compressed=False`. Of course then the output should be adjusted appropriately.
```suggestion
    >>> for index, x in np.ma.ndenumerate(a, compressed=False):
```

**Comment by mattip** [comment]
Thanks for the change in the example. This looks good to me, I will leave it for a little while in case anyone else wants to take a look.

**Comment by mattip** [comment]
Thanks @joukewitteveen

**Comment by InessaPawson** [comment]
Hi-five on merging your first pull request to NumPy, @joukewitteveen! We hope you stick around! Your choices aren’t limited to programming – you can review pull requests, help us stay on top of new and old  issues, develop educational material, work on our website, add or improve graphic design, create marketing materials, translate website content, write grant proposals, and help with other fundraising initiatives. For more info, check out: https://numpy.org/contribute
Also, consider joining our mailing list. This is a great way to connect with other cool people in our community and be part of important conversations that affect the development of NumPy: https://mail.python.org/mailman/listinfo/numpy-discussion

**Pull request merged**